Fix modtime tests on APFS: allow 1us difference

wojas opened the pull request:
Fixes #1087.

This fixes the tests failing on APFS as reported in #1087  by allowing a difference in modtime of up to 1 microsecond.

I'm not sure if this is the right approach. Maybe we should increase the precision that we store in backups instead?

fd0 commented:
So, the underlying issue (as far as I can see) is that the JSON representation is more coarse than what APFS can. So, is there a way to increase the resolution in the JSON representation in a backwards-compatible way?

wojas commented:
This page describes support for nanosecond precision in ISO timestamps across languages:
https://nbsoftsolutions.com/blog/iso-8601-and-nanosecond-precision-across-languages

The reason the JSON timestamp usually do not use the full resolution appears to be that many other languages would not be able to parse it.

Interestingly enough, it looks like the timestamps in Restic's JSON repo objects do use nanosecond resolution: `"mtime": "2017-10-26T03:57:51.192561737+08:00"`. So it appears like something else is limiting the resolution. Maybe the JSON loader?

fd0 commented:
I think the way forward is to find that out. I don't particularly care about what other JSON parsers do, as long as restic is able to use it.

wojas commented:
From syscall/syscall_bsd.go:

```go
func UtimesNano(path string, ts []Timespec) error {
	// TODO: The BSDs can do utimensat with SYS_UTIMENSAT but it
	// isn't supported by darwin so this uses utimes instead
       ...
```

`man utimensat` does return a man page on High Sierra, so maybe `utimensat` was added recently.

The function used in syscall/timestruct.go to round the value:

```go
// NsecToTimeval takes a number of nanoseconds since the Unix epoch
// and returns the corresponding Timeval value.
func NsecToTimeval(nsec int64) Timeval {
	nsec += 999 // round up to microsecond
	usec := nsec % 1e9 / 1e3
	sec := nsec / 1e9
	if usec < 0 {
		usec += 1e6
		sec--
	}
	return setTimeval(sec, usec)
}
```

wojas commented:
`sys/stat.h` confirms this call was added in High Sierra:

```c
int futimens(int __fd, const struct timespec __times[2]) __API_AVAILABLE(macosx(10.13), ios(11.0), tvos(11.0), watchos(4.0));
```

I guess this needs a fix in Go syscall to detect at runtime if this syscall is available. 

The Linux version does something similar:

```go
func UtimesNano(path string, ts []Timespec) (err error) {
        ...
	err = utimensat(_AT_FDCWD, path, (*[2]Timespec)(unsafe.Pointer(&ts[0])))
	if err != ENOSYS {
		return err
	}
	// If the utimensat syscall isn't available (utimensat was added to Linux
	// in 2.6.22, Released, 8 July 2007) then fall back to utimes
        ...
```

I will try to implement a fix in syscall. But even after this is fixed in syscall, it will only be available for future versions of Go, so I guess we still need a workaround in the tests. Maybe only apply this for BSD, and only if the last digits in one of the timestamps are 000?

fd0 commented:
Let's add the fix if `runtime.GOOS() == "darwin"` or so, so that it passes on darwin for now.

wojas committed:
Darwin test fix: allow 1μs timestamp difference

On Darwin, allow a 1μs difference in restored timestamps, because
macOS <10.13 cannot restore with nanosecond precision and the current
version of Go (1.9.2) does not yet support the new syscall required
for this. (#1087 #1389)

wojas commented:
@fd0 Ok, I updated this PR to only apply to darwin.

fd0 reviewed (approved):
LGTM
